### PR TITLE
UGENE-7336 Export the result sequence with the backbone sequence on double click

### PR DESCRIPTION
**Pull request by rasputinkirill**

### CODE REVIEW

File: src/plugins/pcr_primer_design_for_dna_assembly/src/PCRPrimerDesignForDNAAssemblyOPWidget.cpp [hunk 1]
```diff
@@ -348,9 +351,42 @@ void PCRPrimerDesignForDNAAssemblyOPWidget::sl_onFindTaskFinished() {
...
     productsTable->setCurrentProducts(pcrTask->getResults(), annDnaView);
     createResultAnnotations();
+    backboneSequence = pcrTask->getBackboneSequence();
+    lastRunSettings = pcrTask->getSettings();
     pcrTask = nullptr;
 }
 
+void PCRPrimerDesignForDNAAssemblyOPWidget::sl_extractProduct() {
+    ADVSequenceObjectContext *sequenceContext = annDnaView->getActiveSequenceContext();
+    SAFE_POINT(NULL != sequenceContext, L10N::nullPointerError("Sequence Context"), );
```
Comment (DmitriiSukhomlinov) [inline src/plugins/pcr_primer_design_for_dna_assembly/src/PCRPrimerDesignForDNAAssemblyOPWidget.cpp]: `nullptr`

File: src/plugins/pcr_primer_design_for_dna_assembly/src/PCRPrimerDesignForDNAAssemblyOPWidget.cpp [hunk 2]
```diff
@@ -348,9 +351,42 @@ void PCRPrimerDesignForDNAAssemblyOPWidget::sl_onFindTaskFinished() {
...
+    U2SequenceObject *sequenceObject = sequenceContext->getSequenceObject();
+    SAFE_POINT(NULL != sequenceObject, L10N::nullPointerError("Sequence Object"), );
```
Comment (DmitriiSukhomlinov) [inline src/plugins/pcr_primer_design_for_dna_assembly/src/PCRPrimerDesignForDNAAssemblyOPWidget.cpp]: `nullptr`

**Comment by DmitriiSukhomlinov** [comment]
@yalgaer see review

File: src/plugins/pcr_primer_design_for_dna_assembly/src/PCRPrimerDesignForDNAAssemblyOPWidget.cpp [hunk 3]
```diff
@@ -348,9 +351,42 @@ void PCRPrimerDesignForDNAAssemblyOPWidget::sl_onFindTaskFinished() {
...
+    if (sequenceURL.getType() == GUrl_File) {
+        settings.originalSequenceFileName = QFileInfo(sequenceURL.getURLString()).baseName();
+    } else {
+        settings.originalSequenceFileName = sequenceContext->getSequenceObject()->getSequenceName();
+    }
+    if (lastRunSettings.insertTo == PCRPrimerDesignForDNAAssemblyTaskSettings::BackboneBearings::Backbone5) {
+        settings.direction = ExtractPrimerTaskSettings::FiveToFive;
+        settings.backboneLength = lastRunSettings.bachbone5Length;
+    } else {
+        CHECK(lastRunSettings.insertTo == PCRPrimerDesignForDNAAssemblyTaskSettings::BackboneBearings::Backbone3, )
```
Comment (DmitriiSukhomlinov) [inline src/plugins/pcr_primer_design_for_dna_assembly/src/PCRPrimerDesignForDNAAssemblyOPWidget.cpp]: Hmmm... Don't get why this check is needed. Please, add a comment

File: src/plugins/pcr_primer_design_for_dna_assembly/src/PCRPrimerDesignForDNAAssemblyOPWidget.cpp [hunk 4]
```diff
@@ -348,9 +351,42 @@ void PCRPrimerDesignForDNAAssemblyOPWidget::sl_onFindTaskFinished() {
...
+        settings.direction = ExtractPrimerTaskSettings::FiveToThree;
+        settings.backboneLength = lastRunSettings.bachbone3Length;
+    }
+    settings.backboneSequence = backboneSequence;
+    Annotation *selectedAnnotation = productsTable->getSelectedAnnotation();
+    if (selectedAnnotation != nullptr) {
+        CHECK(selectedAnnotation->getRegions().size() == 1, );
```
Comment (DmitriiSukhomlinov) [inline src/plugins/pcr_primer_design_for_dna_assembly/src/PCRPrimerDesignForDNAAssemblyOPWidget.cpp:376]: CHECK of SAFE_POINT?

File: src/plugins/pcr_primer_design_for_dna_assembly/src/tasks/ExtractPrimerTask.cpp [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+    QList<SharedAnnotationData> annotations;
+    int resultIndex = PCRPrimerDesignForDNAAssemblyTask::FRAGMENT_INDEX_TO_NAME.indexOf(settings.fragmentName);
+    
+    U2Region fragmentRegion = U2Region(0, productSequence.length());
+    if (!settings.backboneSequence.isEmpty() && settings.backboneLength != 0) {
+        SharedAnnotationData backboneAnnotationData(new AnnotationData());
+        QByteArray partToAdd = settings.backboneSequence.left(settings.backboneLength);
+        backboneAnnotationData->setStrand(U2Strand(U2Strand::Direct));
+        backboneAnnotationData->name = BACKBONE_ANNOTATION_NAME;
+        if (resultIndex % 2 == 0) {
```
Comment (DmitriiSukhomlinov) [inline src/plugins/pcr_primer_design_for_dna_assembly/src/tasks/ExtractPrimerTask.cpp]: Move `resultIndex % 2` to the separate variable like `bool isForward`

File: src/plugins/pcr_primer_design_for_dna_assembly/src/tasks/ExtractPrimerTask.cpp [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+    AnnotationTableObject *annotationsTableObject = new AnnotationTableObject(productSequence.getName() + " annotations", dbiRef);
+    doc->addObject(annotationsTableObject);
+    annotationsTableObject->addObjectRelation(sequenceObject, ObjectRole_Sequence);
+    annotationsTableObject->addAnnotations(annotations, RESULT_ANNOTATION_GROUP_NAME);
+    result = doc.take();
+}
+
+Document *ExtractPrimerTask::takeResult() {
+    CHECK(nullptr != result, nullptr);
+    if (result->thread() != QCoreApplication::instance()->thread()) {
```
Comment (DmitriiSukhomlinov) [inline src/plugins/pcr_primer_design_for_dna_assembly/src/tasks/ExtractPrimerTask.cpp:121]: The if` statement? You are not sure either the result in the main thread or not?

File: src/plugins/pcr_primer_design_for_dna_assembly/src/tasks/ExtractPrimerTask.h [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+    U2Region fragmentLocation;
+    QString fragmentName;
+    QString originalSequenceFileName;
+    QString outputFileUrl;
+    QByteArray backboneSequence;
+    int backboneLength = 0;
+    BackboneInsertDirection direction = FiveToFive;
+};
+
+class ExtractPrimerTask : public Task {
```
Comment (DmitriiSukhomlinov) [inline src/plugins/pcr_primer_design_for_dna_assembly/src/tasks/ExtractPrimerTask.h:47]: I don't want to criticize your solution but, these two tasks look like it's possible to unit them into one task...

Comment (DmitriiSukhomlinov) [reply src/plugins/pcr_primer_design_for_dna_assembly/src/tasks/ExtractPrimerTask.h:47]: At least split them into separate files

Comment (rasputinkirill) [reply src/plugins/pcr_primer_design_for_dna_assembly/src/tasks/ExtractPrimerTask.h:47]: This is for WD worker.

File: src/plugins/pcr_primer_design_for_dna_assembly/src/tasks/ExtractPrimerTask.cpp (hunk 6 above)
Comment (rasputinkirill) [reply src/plugins/pcr_primer_design_for_dna_assembly/src/tasks/ExtractPrimerTask.cpp:121]: This is for WD worker.

File: src/plugins/pcr_primer_design_for_dna_assembly/src/PCRPrimerDesignForDNAAssemblyOPWidget.cpp (hunk 3 above)
Comment (rasputinkirill) [reply src/plugins/pcr_primer_design_for_dna_assembly/src/PCRPrimerDesignForDNAAssemblyOPWidget.cpp]: Just to be sure what one of options is selected. I can remove this if you found it useless.

File: src/plugins/pcr_primer_design_for_dna_assembly/src/PCRPrimerDesignForDNAAssemblyOPWidget.cpp (hunk 4 above)
Comment (rasputinkirill) [reply src/plugins/pcr_primer_design_for_dna_assembly/src/PCRPrimerDesignForDNAAssemblyOPWidget.cpp:376]: As i can see, it's CHECK. Better to tell me what you want here.

Comment (rasputinkirill) [reply src/plugins/pcr_primer_design_for_dna_assembly/src/PCRPrimerDesignForDNAAssemblyOPWidget.cpp:376]: As i understand CHECK is warning for developer, but SAFE_POINT is for user. And this verification definitely is not for user.

Comment (DmitriiSukhomlinov) [reply src/plugins/pcr_primer_design_for_dna_assembly/src/PCRPrimerDesignForDNAAssemblyOPWidget.cpp:376]: `CHECK` is a legal situation, which could really happen.
`SAFE_POINT` is a kind of situations, which shouldn't happen. 

So, if the condition could fail from time to time and it wont break the application - it's `CHECH`. If the condition should ALWAYS be true  and couldn't be `false` - it's `SAFE_POINT`

File: src/plugins/pcr_primer_design_for_dna_assembly/src/PCRPrimerDesignForDNAAssemblyOPWidget.cpp [hunk 8]
```diff
@@ -348,9 +351,42 @@ void PCRPrimerDesignForDNAAssemblyOPWidget::sl_onFindTaskFinished() {
...
+    settings.sequenceRef = sequenceContext->getSequenceRef();
+    GUrl sequenceURL = sequenceContext->getSequenceObject()->getDocument()->getURL();
+    if (sequenceURL.getType() == GUrl_File) {
+        settings.originalSequenceFileName = QFileInfo(sequenceURL.getURLString()).baseName();
+    } else {
+        settings.originalSequenceFileName = sequenceContext->getSequenceObject()->getSequenceName();
+    }
+    if (lastRunSettings.insertTo == PCRPrimerDesignForDNAAssemblyTaskSettings::BackboneBearings::Backbone5) {
+        settings.direction = ExtractPrimerTaskSettings::FiveToFive;
+        settings.backboneLength = lastRunSettings.bachbone5Length;
```
Comment (DmitriiSukhomlinov) [inline src/plugins/pcr_primer_design_for_dna_assembly/src/PCRPrimerDesignForDNAAssemblyOPWidget.cpp]: You don't need to know this length, you need to add the all sequence!

**Review by DmitriiSukhomlinov — CHANGES_REQUESTED** [review]
The whole backbone sequence should be added to the primer, not only 5' or 3' end!

**Comment by DmitriiSukhomlinov** [comment]
In the following scenarios [gfp.fa](https://ugene.dev/tracker/secure/attachment/19958/gfp.fa) is a target sequence and [backbone.fa](https://ugene.dev/tracker/secure/attachment/19957/backbone.fa) is a backbone sequence.  

Scenario 1:
1. Open the target sequence and set the backbone sequence in the "PCR Primer design"tab.
2. Set minimum "Melting point" to 51.
3. Set "5' insert to 5' backbone" checked.
4. Click "Start".
Expected: "A Forward" and "A Reverse" have been found.
5. Double click "A Forward".
Expected: backbone - left **reverse**, "A Forward" - right forward.
Current: backbone - left **forward**, "A Forward" - right forward.  

Scenario 2:
1. -//-
2. -//-
3. Set "5' insert to 3' backbone" checked.
4. -//-
Expected: "A Forward" and "A Reverse" have been found.
5. Double click "A Reverse".
Expected: "A Reverse" - left reverse, backbone - right **reverse**.
Current: "A Reverse" - left reverse, backbone - right **forward**.

File: src/plugins/pcr_primer_design_for_dna_assembly/src/tasks/ExtractPrimerTask.h [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+#include <U2Core/U2Region.h>
+#include <U2Core/U2Type.h>
+
+namespace U2 {
+
+class Document;
+class LoadDocumentTask;
+
+struct ExtractPrimerTaskSettings {
+    enum BackboneInsertDirection {
```
Comment (DmitriiSukhomlinov) [inline src/plugins/pcr_primer_design_for_dna_assembly/src/tasks/ExtractPrimerTask.h]: Do you really need this extra `enum`? I suppose, `PCRPrimerDesignForDNAAssemblyTaskSettings::BackboneBearings` fits good here and has the same meaning

**Review by DmitriiSukhomlinov — CHANGES_REQUESTED** [review]
There are two fail scenarious

File: src/plugins/pcr_primer_design_for_dna_assembly/src/tasks/ExtractPrimerTask.cpp [hunk 10]
```diff
@@ -0,0 +1,194 @@
...
+    U2SequenceDbi *sequenceDbi = connection.dbi->getSequenceDbi();
+    SAFE_POINT_EXT(NULL != sequenceDbi, setError(L10N::nullPointerError("Sequence DBI")), primerSequence);
+
+    U2Sequence sequence = sequenceDbi->getSequenceObject(settings.sequenceRef.entityId, stateInfo);
+    CHECK_OP(stateInfo, primerSequence);
+
+    primerSequence.seq = sequenceDbi->getSequenceData(settings.sequenceRef.entityId, settings.fragmentLocation, stateInfo);
+    CHECK_OP(stateInfo, primerSequence);
+    primerSequence.setName(settings.fragmentName + " annotations");
+    primerSequence.seq.mid(settings.fragmentLocation.startPos, settings.fragmentLocation.length);
```
Comment (evelinabis) [inline src/plugins/pcr_primer_design_for_dna_assembly/src/tasks/ExtractPrimerTask.cpp:143]: What for?

**Review by DmitriiSukhomlinov — APPROVED** [review]

**Comment by DmitriiSukhomlinov** [comment]
@rasputinkirill please, don't merge until @yalgaer have seen it

**Pull request merged**